static/js/buddy_list.js

```diff
@@ -45,74 +104,195 @@ var buddy_list = (function () {
     };
 
     self.first_key = function () {
-        var list_items = self.container.find(self.item_sel);
-        var li = list_items.first();
-        if (li.length === 0) {
-            return;
-        }
-        var key = self.get_key_from_li({li: li});
-        return key;
+        return self.keys[0];
     };
 
     self.prev_key = function (key) {
-        var li = self.find_li({key: key});
-        var prev_li = li.prev();
-        if (prev_li.length === 0) {
+        var i = self.keys.indexOf(key.toString());
+
+        if (i <= 0) {
             return;
         }
-        var prev_key = self.get_key_from_li({li: prev_li});
-        return prev_key;
+
+        return self.keys[i - 1];
     };
 
     self.next_key = function (key) {
-        var li = self.find_li({key: key});
-        var next_li = li.next();
-        if (next_li.length === 0) {
+        var i = self.keys.indexOf(key.toString());
+
+        if (i < 0) {
             return;
         }
-        var next_key = self.get_key_from_li({li: next_li});
-        return next_key;
+
+        return self.keys[i + 1];
     };
 
     self.maybe_remove_key = function (opts) {
-        var li = self.find_li({key: opts.key});
-        li.remove();
+        var pos = self.keys.indexOf(opts.key);
+
+        if (pos < 0) {
+            return;
+        }
+
+        self.keys.splice(pos, 1);
+
+        if (pos < self.render_count) {
+            self.render_count -= 1;
+            var li = self.find_li({key: opts.key});
+            li.remove();
+        }
     };
 
-    self.insert_or_move = function (opts) {
+    self.find_position = function (opts) {
         var key = opts.key;
+        var i;
+
+        for (i = 0; i < self.keys.length; i += 1) {
```

Comment:
Trying to take the time to read through this without much WebApp experience. Looks like a ton of work! 

Anyway, just wanted to throw a quick heads up that as soon as we have babel integrated we'll have access to a TON of newer ECMAScript 5 and 6 features that make writing this kind of JS much more fun and compatible even with older browsers like IE9 and under. Like [Array.prototype.forEach](https://gomakethings.com/looping-through-arrays-the-es6-way). 

Note: This may have been more of a chat thing but I just wanted to mention it :)

Reply:
We have been using `_.each(...)` for several years in our codebase, and we're pretty happy with that.  My reason for using a `for` loop here is that I'm searching for the first value of `i` meeting a condition, and I want to early-exit.  I tend to use `each` more in situations where you're producing a side effect for every element and `map` when I'm creating a new array.

Reply:
My mistake I guess I didn't really take the time to read the function carefully enough. In any case I know _.each does a bunch of performance optimization so its not a bad choice at all. I'm a huge fan of j dalton and his work. Glad to hear that!

Reply:
I also prefer lodash to underscore. both are essentially the same the benefit of switching to lodash is as @armaanahluwalia mentioned allows us to only bundle part of the code that needed rather than the whole library. Shipping whole underscore or lodash just increases the js load on browser.

Reply:
Can we move the lodash-vs.-underscore kibitzing to a new issue or discuss on chat?  It's not really relevant to this particular PR.  Feel free to link the issue here to close out that part of the conversation.

Reply:
Sure. Its a step 2 thing after we actually have modules in place. Not so relevant at the moment. Sorry for sidetracking.